### PR TITLE
Issue #1039: Use browser-engine-system and feature-session components to replace WebView abstraction layer..

### PR DESCRIPTION
**Pull request by pocmo**
@mcomella @Baron-Severin I want to share the current state of the migration to use `browser-engine`. This version is functional and you can use it to browse the web. Like with the `browser-session` integration I used extension functions to work around functionality that is not available in the upstream components.

Missing pieces without a workaround in this version (list may be incomplete):
* **Tracking protection** toggle not functional (TP is already in components 0.20, with 0.21 [Friday] we get the option to configure this globally for an engine or session and that will make this toggle simpler to implement.
* We can't **intercept URLs** to inject HTML currently (used for firefox:home, firefox:about and error pages). ~~This functionality is scheduled to land in 0.22 (next week). However I may move this into the current sprint so that we can have this in 0.21 on Friday.~~ We moved this into the 0.21 sprint and the functionality should be available in 0.21.
  * Without loading "firefox:home" and adding it to the tab's history this currently breaks some assumptions when using the back button. This should fix itself once "firefox:home" gets loaded normally.
* **Fullscreen API** (works with YouTube because the TV website doesn't use the web API and just shows the content at full size). Scheduled to land in 0.21 on Friday.

Added follow-up items to: #1063 

---

As a side effect:
 * Replaces AmazonWebView with WebView (Issue #379)
 * Removes the product flavor dimensions that are not needed anymore.

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/Components.kt
Comment (mcomella) [inline app/src/main/java/org/mozilla/focus/Components.kt]: nit: by lazy? Otherwise engine and sessionManager will make themselves right now.

**Review by mcomella — COMMENTED** [review]
With a quick skim, this looks reasonable to me. We can spend more time reviewing when we land for real (next week?). Note: I'll be PTO next week so I'll be unable to look over this until the following Monday.

image

You win Fire TV!

**Comment by mcomella** [comment]
> Note: I'll be PTO next week so I'll be unable to look over this until the following Monday.

Unless we can land something tomorrow. :)

File: app/src/main/java/org/mozilla/focus/Components.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/Components.kt]: Oh, yeah, good catch!

**Comment by pocmo** [comment]
> @pocmo How much is changing for the final PR?

For this iteration not much.

Components 0.21 should get released today and then with that I should be able to look at the three missing pieces mentioned above. The patches for tracking protection and URL interception landed in a-c. Fullscreen API is at risk for 0.21 because the PR is not up yet. We may need to ship that with 0.22 or can ship a 0.21.1 if this gets in early next week.

> Provided the changes aren't too big, I can look over the most recent WIP tomorrow, Severin can look at it next week and land it, and I can look at it to make sure we didn't miss anything when I come back. 

Depending on what you and @Baron-Severin think I'd be open to land this as-is (with review comments addressed) and open separate PRs for the next pieces. The idea is to unblock others by having to wait on a big code drop and to be able to have others test and iterate on the new code. The drawback is that we temporarily have a not 100% feature complete app though (-> at least the three points mentioned above). Apart from that the app seems to be "fully" functional in my tests.

**Comment by pocmo** [comment]
Updated the PR. Made all checks pass (ktlint, detekt, checkstyle, lint), SessionUseCases initializes lazily now and cleaned up some of the code.

**Review by mcomella — APPROVED** [review]
A few things missing here or there and a few nits but otherwise lgtm. I find the `SessionUseCases` abstraction strange: see below.

Some other things that have changed – please file a bug or fix:
- Back/forward state takes a while to update

Not sure it was related to your changes but I got into a state where a white page was loaded and the progress bar was shown without a URL (i.e. empty str or null) after going  backwards and forwards a few times.

> I'd be open to land this as-is (with review comments addressed) and open separate PRs for the next pieces.

Me too: please send an email to the team letting them know 1) this is landing, 2) the features that need follow-ups, and 3) that they'll need to change their build variant. Also, don't forget to file issues to track this work! :)

@Baron-Severin FYI that this may land.

File: app/build.gradle
Comment (pocmo) [reply app/build.gradle]: Yeah, now that one dimension is gone they are just "debug" etc..

File: app/src/main/java/org/mozilla/focus/ext/EngineView.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/ext/EngineView.kt]: Makes sense. I'll add that.

File: app/src/amazonWebview/java/org/mozilla/focus/webview/FirefoxAmazonWebView.kt
Comment (pocmo) [reply app/src/amazonWebview/java/org/mozilla/focus/webview/FirefoxAmazonWebView.kt]: Added this to list #1063.

File: app/src/main/java/org/mozilla/focus/ext/EngineView.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/ext/EngineView.kt]: Added this to list in #1063.

File: app/src/main/java/org/mozilla/focus/browser/BrowserFragment.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/browser/BrowserFragment.kt]: I saw you filed https://github.com/mozilla-mobile/android-components/issues/697. Let's discuss there.

**Comment by pocmo** [comment]
From the migration I feel like the overlay and the things touching the engine are pretty well separated.  We are definitely introducing risk with the migration but it doesn't feel like it gets worse by having both things (engine migration + overlay migration) happen at the same time.

> whose components haven't been used in a production app before

Yeah, that's a chicken egg problem. FFFTV is the first candidate because the feature set is the smallest and the first milestone we have reached. :)

> there are regressions in back/forward state that I couldn't reproduce consistently 

I noticed that there is a regression because we can't intercept "firefox:home" and inject custom content. With that the page is not in the history stack and breaks some of the logic. The functionality for that landed in 0.21 - I added an item about testing back/forward after that to #1063.

> Perhaps we should land but not hesitate to back it out if it causes problems? What do you think @pocmo ?

Yeah, that sounds good to me. It will get more complicated to back out later on after we added more commits on top of it though.

I'll talk to QA once this lands. :)

**Comment by pocmo** [comment]
@Baron-Severin From my side this is ready to land. I addressed most of @mcomella's review comments, rebased the branch and added follow-up items to #1063. I started working on the follow-up items now (using 0.21 of the components) but I'll create separate PRs for those changes once this PR lands.

File: app/src/main/java/org/mozilla/focus/ext/Fragment.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/ext/Fragment.kt]: It's not shorter but I think it shows the intent better? If I see `context!!` I'm never sure whether someone just wanted to make the code compile or if they actually *assume* that this will not be null here. Naming-wise I followed `requireContext` and `requireActivity` from Fragment. And I noticed I should use `requireContext` here... :)

File: app/src/main/java/org/mozilla/focus/utils/IntentUtils.java
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/utils/IntentUtils.java]: Oh, I missed that part. I'll update the code.

File: app/src/main/java/org/mozilla/focus/iwebview/EngineViewLifecycleFragment.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/iwebview/EngineViewLifecycleFragment.kt]: Ah, nice, thanks! I'll refactor that. :)

File: app/src/main/java/org/mozilla/focus/ext/EngineView.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/ext/EngineView.kt]: Ah, this shouldn't be necessary anymore. I needed that during the migration when I had not implemented all extension methods yet. :)

File: app/src/main/java/org/mozilla/focus/browser/InfoFragment.java
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/browser/InfoFragment.java]: Added to #1063.

File: app/src/main/java/org/mozilla/focus/browser/InfoFragment.java
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/browser/InfoFragment.java]: Added to #1063.

File: app/src/main/java/org/mozilla/focus/browser/InfoActivity.java
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/browser/InfoActivity.java]: Added to #1063.

**Review by severinrudie — APPROVED** [review]
Had a few questions, nothing major.  LGTM

File: app/build.gradle
Comment (pocmo) [reply app/build.gradle]: Previously we had product flavors for the engine and those flavors added a suffix:

```Groovy
roductFlavors {
        amazonWebview {
            applicationIdSuffix ".firefox"
            // [..]
        }
        gecko {
            applicationIdSuffix ".gecko"
            // [..]
        }
    }
```

We have been shipping the `amazonWebview` flavor only. Now that the flavor is gone I moved the suffix to the `applicationId` setup directly.

File: app/src/main/java/org/mozilla/focus/browser/InfoActivity.kt
Comment (pocmo) [reply app/src/main/java/org/mozilla/focus/browser/InfoActivity.kt]: That's a good idea and avoids some repetition. I'll update the PR.

**Pull request merged**
